dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -401,21 +405,26 @@ public UfsUrl join(String suffix) {
     while (nonEmptyIndex < suffixArray.length && suffixArray[nonEmptyIndex].isEmpty())  {
       nonEmptyIndex++;
     }
-    List<String> suffixComponentsList;
+    List<String> suffixComponentsList = Arrays.asList(suffixArray);
+    List<String> noEmptyElemSuffixComponentsList;
     if (nonEmptyIndex == 0) {
-      suffixComponentsList = Arrays.asList(suffixArray);
+      noEmptyElemSuffixComponentsList = suffixComponentsList;
     } else {
-      // TODO(Tony Sun): optimize the performance later
-      suffixComponentsList = Arrays.asList(
-          Arrays.copyOfRange(
-              suffixArray,
-              nonEmptyIndex, suffixArray.length));
+      noEmptyElemSuffixComponentsList = suffixComponentsList.subList(nonEmptyIndex,
+          suffixComponentsList.size());
     }
     List<String> pathComponents = mProto.getPathComponentsList();
     return new UfsUrl(UfsUrlMessage.newBuilder()
         .setScheme(mProto.getScheme())
         .setAuthority(mProto.getAuthority())
         .addAllPathComponents(pathComponents)
-        .addAllPathComponents(suffixComponentsList).build());
+        .addAllPathComponents(noEmptyElemSuffixComponentsList).build());
+  }
+
+  public UfsUrl joinUnsafe(String suffix) {
+    String path = getFullPath();
+    StringBuilder sb = new StringBuilder(path.length() + 1 + suffix.length());
+    return new UfsUrl(toProto().getScheme(), toProto().getAuthority(),
+        sb.append(path).append(SLASH_SEPARATOR).append(suffix).toString());
```

Comment:
this can call `UfsUrl(UfsUrlMessage msg)` to avoid the cost to build a new path string and then immediately after parse into a path components list.

```suggestion
UfsUrlMessage urlWithSuffix = UfsUrlMessage.newBuilder().mergeFrom(this.mProto).build();
return new UfsUrl(urlWithSuffix);
```

Reply:
The Code here has been changed.